README.md
# Lockit signup

[![Build Status](https://travis-ci.org/zemirco/lockit-signup.svg?branch=master)](https://travis-ci.org/zemirco/lockit-signup)
[![NPM version](https://badge.fury.io/js/lockit-signup.svg)](http://badge.fury.io/js/lockit-signup)
[![Dependency Status](https://david-dm.org/zemirco/lockit-signup.svg)](https://david-dm.org/zemirco/lockit-signup)

Sign up users to your Express app. The module is part of [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-signup`

```js
var Signup = require('lockit-signup');
var lockitUtils = require('lockit-utils');
var config = require('./config.js');

var db = lockitUtils.getDatabase(config);
var adapter = require(db.adapter)(config);

var app = express();

// express settings
// ...
// sessions are required - either cookie or some sort of db
app.use(cookieParser());
app.use(cookieSession({
  secret: 'this is my super secret string'
}));

// create new Signup instance
var signup = new Signup(config, adapter);

// use signup.router with your app
app.use(signup.router);
```

## Configuration

More about configuration at [lockit](https://github.com/zemirco/lockit).

## Features

 - validate inputs
 - hash password
 - validation link expiration
 - verify email address via unique tokens
 - prevent duplicate email/username sign up
 - resend verification email

## Routes included

 - `GET /signup`
 - `POST /signup`
 - `GET /signup/:token`
 - `GET /signup/resend-verification`
 - `POST /signup/resend-verification`

## REST API

If you've set `exports.rest = true` in your `config.js` the module behaves as follows.

 - all routes have `/rest` prepended
 - `GET /rest/signup` is `next()`ed and you can catch `/signup` on the client
 - `POST /rest/signup` stays the same but only sends JSON
 - `GET /rest/signup/:token` sends JSON
 - `GET /rest/signup/resend-verification` is `next()`ed and you can catch `/signup/resend-verification` on the client
 - `POST /rest/signup/resend-verification` sends JSON

## Test

`grunt`

## License

MIT


History.md

##### 1.2.2 / 2015-07-06

- update dependencies

##### 1.2.1 / 2015-07-03

- update dependencies
- use eslint
- use make instead grunt

##### 1.2.0 / 2014-09-27

- update dependencies

##### 1.1.1 / 2014-08-18

- keep username and email field values on failed signup request
- update dependencies

##### 1.1.0 / 2014-05-27

- improve views by using Bootstrap conditional classes
- set `autocomplete="off"` for forms
- refactor code
- use updated `lockit-sendmail` module
- update dependencies

##### 1.0.0 / 2014-04-19

- requires Express 4.x
- makes use of `express.Router()`. No need to pass `app` around as argument.

  **old**

  ```js
  var Signup = require('lockit-signup');

  var signup = new Signup(app, config, adapter);
  ```

  **new**

  ```js
  var Signup = require('lockit-signup');

  var signup = new Signup(config, adapter);
  app.use(signup.router);
  ```

- proper Error handling. All Errors are piped to next middleware.

  **old**

  ```js
  if (err) console.log(err);
  ```

  **new**

  ```js
  if (err) return next(err);
  ```

  Make sure you have some sort of error handling middleware at the end of your
  routes (is included by default in Express 4.x apps if you use the `express-generator`).


##### 0.7.1 / 2014-04-14

- username (`name`) has to be lowercase and has to start with a letter

##### 0.7.0 / 2014-04-11

- `username` becomes `name`


test/app.js

/**
 * Module dependencies.
 */

var express = require('express');
var routes = require('./routes');
var user = require('./routes/user');
var http = require('http');
var path = require('path');

var signup = require('../index.js');

function start(config) {

  config = config || require('./config.js');

  var app = express();

  // set basedir so views can properly extend layout.jade
  app.locals.basedir = __dirname + '/views';

  // all environments
  app.set('port', process.env.PORT || 3000);
  app.set('views', __dirname + '/views');
  app.set('view engine', 'jade');
  // make JSON output simpler for testing
  app.set('json spaces', 0);
  app.use(express.favicon());
  app.use(express.bodyParser());
  app.use(express.methodOverride());
  app.use(express.cookieParser('your secret here'));
  app.use(express.cookieSession());
  
  // testing csrf
  if (config.csrf) {
    app.use(express.csrf());
    app.use(function(req, res, next) {
      
      var token = req.csrfToken();
      res.locals._csrf = token;
      
      // save token to a cookie so we can easily access it on the client
      res.cookie('csrf', token);
      next();
    });
  }

  // use forgot password middleware with testing options
  signup(app, config);

  app.use(app.router);
  app.use(express.static(path.join(__dirname, 'public')));

  // development only
  if ('development' == app.get('env')) {
    app.use(express.errorHandler());
  }

  app.get('/', routes.index);
  app.get('/users', user.list);

  http.createServer(app).listen(app.get('port'), function(){
    console.log('Express server listening on port ' + app.get('port'));
  });

  return app;

}

// export app for testing
if(require.main === module){
  // called directly
  start();
} else {
  // required as a module -> from test file
  module.exports = function(config) {
    return start(config);
  };
}

test/csrf.js

var request = require('supertest');
var should = require('should');
var cookie = require('cookie');

var config = require('./app/config.js');
var app = require('./app/app.js');
var adapter = require('lockit-couchdb-adapter')(config);

// clone config object
var _config = JSON.parse(JSON.stringify(config));
_config.port = 8000;
_config.csrf = true;
_config.signup.tokenExpiration = '1 ms';

var _app = app(_config);

describe('# with csrf', function() {

  describe('GET /signup', function() {
    it('should include the token in the view', function(done) {
      request(_app)
          .get('/signup')
          .end(function(err, res) {
            var cookies = cookie.parse(res.headers['set-cookie'][0]);
            var token = cookies.csrf;
            res.text.should.containEql('name="_csrf" value="' + token + '"');
            done();
          });
    });
  });

  describe('GET /signup/resend-verification', function() {
    it('should include the token in the view', function(done) {
      request(_app)
          .get('/signup/resend-verification')
          .end(function(err, res) {
            var cookies = cookie.parse(res.headers['set-cookie'][0]);
            var token = cookies.csrf;
            res.text.should.containEql('name="_csrf" value="' + token + '"');
            done();
          });
    });
  });

  describe('GET /signup/:token', function() {
    it('should include the token in the view', function(done) {

      // need to create another adapter so token expires immediately
      var _adapter = require('lockit-couchdb-adapter')(_config);

      _adapter.save('csrf', 'csrf@email.com', 'pass', function() {
        _adapter.find('name', 'csrf', function(err, user) {
          request(_app)
              .get('/signup/' + user.signupToken)
              .end(function(err, res) {
                var cookies = cookie.parse(res.headers['set-cookie'][0]);
                var token = cookies.csrf;
                res.text.should.containEql('name="_csrf" value="' + token + '"');
                done();
              });
        });
      });
    });
  });

  after(function(done) {
    adapter.remove('csrf', done);
  });

});


test/rest.js

var request = require('supertest');
var should = require('should');

var config = require('./app/config.js');
var app = require('./app/app.js');
var adapter = require('lockit-couchdb-adapter')(config);

// create an app with rest enabled
var _config_one = JSON.parse(JSON.stringify(config));
_config_one.port = 6000;
_config_one.rest = true;
_config_one.signup.tokenExpiration = '10 ms';

// create a second app with longer token expiration
var _config_two = JSON.parse(JSON.stringify(_config_one));
_config_two.port = 6001;
_config_two.signup.tokenExpiration = '1 h';

var _app_one = app(_config_one);
var _app_two = app(_config_two);

describe('# with REST enabled', function() {

  describe('GET /signup', function() {

    it('should not handle the route when REST is active', function(done) {
      request(_app_one)
        .get('/rest/login')
        .end(function(error, res) {
          res.statusCode.should.equal(404);
          done();
        });
    });

  });

  describe('POST /signup', function() {

    it('should return an error when one input is blank (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup')
        .send({name: '', email: 'some@email.com', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"All fields are required"}');
          done();
        });
    });

    it('should return an error when username contains non-url-safe chars (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup')
        .send({name: 'name@', email: 'some@email.com', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Username may not contain any non-url-safe characters"}');
          done();
        });
    });

    it('should return an error when email has invalid format (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup')
        .send({name: 'some', email: 'someemail.com', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Email is invalid"}');
          done();
        });
    });

    it('should render a success message when everything went fine (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup')
        .send({name: 'steve', email: 'steve@wayne.com', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

    it('should return an error message username is already taken (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup')
        .send({name: 'steve', email: 'steve@wayne.com', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Username already taken"}');
          done();
        });
    });

    it('should render a success message when duplicate email was found (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup')
        .send({name: 'jeff', email: 'steve@wayne.com', password: 'secret'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

  });

  describe('GET /signup/:token', function() {

    it('should render 404 message when token is invalid (REST)', function(done) {
      request(_app_one)
        .get('/rest/signup/id123')
        .end(function(error, res) {
          res.statusCode.should.equal(404);
          done();
        });
    });

    it('should render an error message when signup token has expired (REST)', function(done) {

      // first sign up a new user -> beep
      request(_app_one)
        .post('/rest/signup')
        .send({name: 'beep', email: 'beep@wayne.com', password: 'secret'})
        .end(function(err, res) {

          // second get beep's signup token
          adapter.find('name', 'beep', function(err, user) {

            // third call url with token
            request(_app_one)
              .get('/rest/signup/' + user.signupToken)
              .end(function(error, res) {
                res.statusCode.should.equal(403);
                res.text.should.equal('{"error":"token expired"}');
                done();
              });

          });

        });

    });

    it('should render a success message when token is valid (REST)', function(done) {

      request(_app_two)
        .post('/rest/signup')
        .send({name: 'steward', email: 'steward@wayne.com', password: 'secret'})
        .end(function(error, res) {

          // get token for our test user 'steward'
          adapter.find('name', 'steward', function(err, user) {

            // request url with token
            request(_app_one)
              .get('/rest/signup/' + user.signupToken)
              .end(function(error, res) {
                res.statusCode.should.equal(204);
                done();
              });

          });

        });

    });

  });

  describe('GET /signup/resend-verification', function() {

    it('should not catch the route when REST is active', function(done) {
      request(_app_one)
        .get('/rest/signup/resend-verification')
        .end(function(error, res) {
          res.statusCode.should.equal(404);
          done();
        });
    });

  });

  describe('POST /signup/resend-verification', function() {

    it('should return an error when email has invalid format (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup/resend-verification')
        .send({email: 'somewayne.com'})
        .end(function(error, res) {
          res.statusCode.should.equal(403);
          res.text.should.equal('{"error":"Email is invalid"}');
          done();
        });
    });

    it('should render a success message when no existing user was found (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup/resend-verification')
        .send({email: 'jim@wayne.com'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });

    it('should render a succes message when email address is already verified (REST)', function(done) {
      request(_app_one)
        .post('/rest/signup/resend-verification')
        .send({email: 'steward@wayne.com'})
        .end(function(error, res) {
          res.statusCode.should.equal(204);
          done();
        });
    });


    it('should render a success message when email was sent (REST)', function(done) {
      // create new user but do not activate
      adapter.save('mike', 'mike@email.com', 'pass', function() {
        request(_app_one)
          .post('/rest/signup/resend-verification')
          .send({email: 'mike@email.com'})
          .end(function(error, res) {
            res.statusCode.should.equal(204);
            done();
          });
      });
    });

  });

  after(function(done) {
    adapter.remove('beep', function() {
      adapter.remove('steve', function() {
        adapter.remove('steward', function() {
          adapter.remove('mike', done);
        });
      });
    });
  });

});


index.js
'use strict';

var path = require('path');
var events = require('events');
var util = require('util');
var express = require('express');
var uuid = require('node-uuid');
var ms = require('ms');
var moment = require('moment');
var Mail = require('lockit-sendmail');



/**
 * Internal helper functions
 */
function join(view) {
  return path.join(__dirname, 'views', view);
}



/**
 * Signup constructor function.
 *
 * @constructor
 * @param {Object} config
 * @param {Object} adapter
 */
var Signup = module.exports = function(config, adapter) {
  if (!(this instanceof Signup)) {return new Signup(config, adapter); }
  events.EventEmitter.call(this);

  this.config = config;
  this.adapter = adapter;

  var route = config.signup.route || '/signup';
  if (config.rest) {route = '/rest' + route; }

  var router = new express.Router();
  router.get(route, this.getSignup.bind(this));
  router.post(route, this.postSignup.bind(this));
  router.get(route + '/resend-verification', this.getSignupResend.bind(this));
  router.post(route + '/resend-verification', this.postSignupResend.bind(this));
  router.get(route + '/:token', this.getSignupToken.bind(this));
  this.router = router;
};

util.inherits(Signup, events.EventEmitter);



/**
 * GET /signup.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Signup.prototype.getSignup = function(req, res, next) {
  // do not handle the route when REST is active
  if (this.config.rest) {return next(); }

  // custom or built-in view
  var view = this.config.signup.views.signup || join('get-signup');

  res.render(view, {
    title: 'Sign up',
    basedir: req.app.get('views')
  });
};



/**
 * POST /signup.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Signup.prototype.postSignup = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;
  var that = this;

  var name = req.body.name;
  var email = req.body.email;
  var password = req.body.password;

  var error = null;
  // regexp from https://github.com/angular/angular.js/blob/master/src/ng/directive/input.js#L4
  var EMAIL_REGEXP = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,6}$/;

  // check for valid inputs
  if (!name || !email || !password) {
    error = 'All fields are required';
  } else if (name !== encodeURIComponent(name)) {
    error = 'Username may not contain any non-url-safe characters';
  } else if (name !== name.toLowerCase()) {
    error = 'Username must be lowercase';
  } else if (!name.charAt(0).match(/[a-z]/)) {
    error = 'Username has to start with a lowercase letter (a-z)';
  } else if (!email.match(EMAIL_REGEXP)) {
    error = 'Email is invalid';
  }

  // custom or built-in view
  var errorView = config.signup.views.signup || join('get-signup');

  if (error) {
    // send only JSON when REST is active
    if (config.rest) {return res.json(403, {error: error}); }

    // render template with error message
    res.status(403);
    return res.render(errorView, {
      title: 'Sign up',
      error: error,
      basedir: req.app.get('views'),
      name: name,
      email: email
    });
  }

  // check for duplicate name
  adapter.find('name', name, function(err, user) {
    if (err) {return next(err); }

    if (user) {
      error = 'Username already taken';
      // send only JSON when REST is active
      if (config.rest) {return res.json(403, {error: error}); }

      // render template with error message
      res.status(403);
      return res.render(errorView, {
        title: 'Sign up',
        error: error,
        basedir: req.app.get('views'),
        name: name,
        email: email
      });
    }

    // check for duplicate email - send reminder when duplicate email is found
    adapter.find('email', email, function(findErr, foundUser) {
      if (findErr) {return next(findErr); }

      // custom or built-in view
      var successView = config.signup.views.signedUp || join('post-signup');

      if (foundUser) {
        // send already registered email
        var mail = new Mail(config);
        return mail.taken(foundUser.name, foundUser.email, function(takenErr) {
          if (takenErr) {return next(takenErr); }

          // send only JSON when REST is active
          if (config.rest) {return res.send(204); }

          res.render(successView, {
            title: 'Sign up - Email sent',
            basedir: req.app.get('views')
          });
        });

      }

      // looks like everything is fine

      // save new user to db
      adapter.save(name, email, password, function(saveErr, savedUser) {
        if (saveErr) {return next(saveErr); }

        // send email with link for address verification
        var m = new Mail(config);
        m.signup(savedUser.name, savedUser.email, savedUser.signupToken, function(signupErr) {
          if (signupErr) {return next(signupErr); }

          // emit event
          that.emit('signup::post', savedUser);

          // send only JSON when REST is active
          if (config.rest) {return res.send(204); }

          res.render(successView, {
            title: 'Sign up - Email sent',
            basedir: req.app.get('views')
          });
        });

      });

    });

  });
};



/**
 * GET /signup/resend-verification.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Signup.prototype.getSignupResend = function(req, res, next) {
  // do not handle the route when REST is active
  if (this.config.rest) {return next(); }

  // custom or built-in view
  var view = this.config.signup.views.resend || join('resend-verification');

  res.render(view, {
    title: 'Resend verification email',
    basedir: req.app.get('views')
  });
};



/**
 * POST /signup/resend-verification.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Signup.prototype.postSignupResend = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;

  var email = req.body.email;

  var error = null;
  // regexp from https://github.com/angular/angular.js/blob/master/src/ng/directive/input.js#L4
  var EMAIL_REGEXP = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,6}$/;

  if (!email || !email.match(EMAIL_REGEXP)) {
    error = 'Email is invalid';
  }

  if (error) {
    // send only JSON when REST is active
    if (config.rest) {return res.json(403, {error: error}); }

    // custom or built-in view
    var errorView = config.signup.views.resend || join('resend-verification');

    // render template with error message
    res.status(403);
    return res.render(errorView, {
      title: 'Resend verification email',
      error: error,
      basedir: req.app.get('views')
    });
  }

  // check for user with given email address
  adapter.find('email', email, function(err, user) {
    if (err) {return next(err); }

    // custom or built-in view
    var successView = config.signup.views.signedUp || join('post-signup');

    // no user with that email address exists -> just render success message
    // or email address is already verified -> user has to use password reset function
    if (!user || user.emailVerified) {
      // send only JSON when REST is active
      if (config.rest) {return res.send(204); }

      return res.render(successView, {
        title: 'Sign up - Email sent',
        basedir: req.app.get('views')
      });
    }

    // we have an existing user with provided email address

    // create new signup token
    var token = uuid.v4();

    // save token on user object
    user.signupToken = token;

    // set new sign up token expiration date
    var timespan = ms(config.signup.tokenExpiration);
    user.signupTokenExpires = moment().add(timespan, 'ms').toDate();

    // save updated user to db
    adapter.update(user, function(updateErr, updatedUser) {
      if (updateErr) {return next(updateErr); }

      // send sign up email
      var mail = new Mail(config);
      mail.resend(updatedUser.name, email, token, function(resendErr) {
        if (resendErr) {return next(resendErr); }

        // send only JSON when REST is active
        if (config.rest) {return res.send(204); }

        res.render(successView, {
          title: 'Sign up - Email sent',
          basedir: req.app.get('views')
        });
      });

    });

  });
};



/**
 * GET /signup/:token.
 *
 * Route is at the end so it does not
 * catch :token === 'resend-verification'.
 *
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
Signup.prototype.getSignupToken = function(req, res, next) {
  var config = this.config;
  var adapter = this.adapter;
  var that = this;

  var token = req.params.token;

  // verify format of token
  var re = new RegExp('[0-9a-f]{22}|[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}', 'i');

  // if format is wrong no need to query the database
  if (!re.test(token)) {return next(); }

  // find user by token
  adapter.find('signupToken', token, function(err, user) {
    if (err) {return next(err); }

    // no user found -> forward to error handling middleware
    if (!user) {return next(); }

    // check if token has expired
    if (new Date(user.signupTokenExpires) < new Date()) {

      // delete old token
      delete user.signupToken;

      // save updated user to db
      return adapter.update(user, function(updateErr) {
        if (updateErr) {return next(updateErr); }

        // send only JSON when REST is active
        if (config.rest) {return res.json(403, {error: 'token expired'}); }

        // custom or built-in view
        var expiredView = config.signup.views.linkExpired || join('link-expired');

        // render template to allow resending verification email
        res.render(expiredView, {
          title: 'Sign up - Email verification link expired',
          basedir: req.app.get('views')
        });

      });

    }

    // everything seems to be fine

    // set user verification values
    user.emailVerificationTimestamp = new Date();
    user.emailVerified = true;

    // remove token and token expiration date from user object
    delete user.signupToken;
    delete user.signupTokenExpires;

    // save user with updated values to db
    adapter.update(user, function(updateErr, updatedUser) {
      if (updateErr) {return next(updateErr); }

      // emit 'signup' event
      that.emit('signup', updatedUser, res);

      if (config.signup.handleResponse) {

        // send only JSON when REST is active
        if (config.rest) {return res.send(204); }

        // custom or built-in view
        var view = config.signup.views.verified || join('mail-verification-success');

        // render email verification success view
        res.render(view, {
          title: 'Sign up success',
          basedir: req.app.get('views')
        });

      }

    });

  });
};
